### PR TITLE
Added learner journey APIs & tests

### PR DESCRIPTION
**Pull request by karan-devslane**
1. Added the following learner journey APIs: Create, Update & Read
2. Added unit tests for the above APIs
3. Added `.nvmrc` file for easy node version switching use nvm

### CODE REVIEW

File: src/controllers/LearnerJourneyCreate/LearnerJourneyCreate.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  if (!isRequestValid.isValid) {
+    const code = 'LEARNER_JOURNEY_INVALID_INPUT';
+    logger.error({ code, apiId, msgid, resmsgid, requestBody, message: isRequestValid.message });
+    throw amlError(code, isRequestValid.message, 'BAD_REQUEST', 400);
+  }
+
+  // TODO: validate question_set_id
+
+  const learnerJourneyInsertData = {
+    ...dataBody,
```
Comment (HarishGangula) [inline src/controllers/LearnerJourneyCreate/LearnerJourneyCreate.ts:34]: validate the request body having , valid  date for start time

File: src/services/learnerJourney.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+};
+
+export const updateLearnerJourney = async (identifier: string, req: UpdateLearnerJourney): Promise<any> => {
+  const whereClause: Record<string, any> = { identifier };
+  const updatedLearnerJourney = await LearnerJourney.update(req, { where: whereClause });
+  return { updatedLearnerJourney };
+};
+
+export const readLearnerJourney = async (learnerId: string): Promise<{ learnerJourney: any }> => {
+  const learnerJourney = await LearnerJourney.findOne({
```
Comment (HarishGangula) [inline src/services/learnerJourney.ts:18]: This works but always get the order by lastest updated time

File: src/routes/entities/learnerRouter.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import express from 'express';
+import { learnerJourneyRouter } from './learnerJourneyRouter';
+
+export const learnerRouter = express.Router({ mergeParams: true });
+
+learnerRouter.use('/:learner_id/journey', learnerJourneyRouter);
```
Comment (HarishGangula) [inline src/routes/entities/learnerRouter.ts:6]: learner_id should be at the end

**Review by HarishGangula — CHANGES_REQUESTED** [review]
Add the open api docs to review the code for the spec

**Comment by karan-devslane** [comment]
Swagger API Doc:

```
openapi: 3.0.0
info:
  title: Learner Journey Data
  version: 1.0.0
servers:
  - url: http://{{base_url}}
components:
  securitySchemes:
    bearerAuth:
      type: http
      scheme: bearer
paths:
  /v1/learner/journey/create/{learner_id}:
    post:
      tags:
        - default
      summary: Create
      requestBody:
        content:
          application/json:
            schema:
              type: object
              example:
                question_set_id: 37a60112-1136-40cc-b968-2f664feb1864
                start_time: '2024-09-03T12:34:56Z'
      security:
        - bearerAuth: []
      parameters:
        - name: msgid
          in: query
          schema:
            type: string
          example: d297e367-7b72-45c1-883c-18005a7ea888
        - name: learner_id
          in: path
          schema:
            type: string
          required: true
          example: 2ca3b11a-d581-4aa1-a199-fb1c51392cc9
      responses:
        '201':
          description: Created
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.create
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: SUCCESS
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: OK
                result:
                  message: Learner journey started successfully
                  identifier: d297e367-7b72-45c1-883c-18005a7ea888
        '400':
          description: Bad Request
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.create
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: BAD_REQUEST
                result: null
                error:
                  code: LEARNER_JOURNEY_INVALID_INPUT
                  message: learner_id must be an integer and is required
        '404':
          description: Not Found
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.create
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: NOT_FOUND
                result: null
                error:
                  code: LEARNER_NOT_FOUND
                  message: Learner with id 10 does not exist
  /v1/learner/journey/update/{learner_id}:
    put:
      tags:
        - default
      summary: Update
      requestBody:
        content:
          application/json:
            schema:
              type: object
              example:
                start_time: '2024-09-03T12:34:56Z'
                end_time: '2024-09-10T12:34:56Z'
                completed_question_ids:
                  - 123e4567-e89b-12d3-a456-426614174000
                status: in_progress
                attempts_count: 2
      security:
        - bearerAuth: []
      parameters:
        - name: msgid
          in: query
          schema:
            type: string
          example: d297e367-7b72-45c1-883c-18005a7ea888
        - name: learner_id
          in: path
          schema:
            type: string
          required: true
          example: 2ca3b11a-d581-4aa1-a199-fb1c51392cc9
      responses:
        '201':
          description: Created
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.update
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: SUCCESS
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: OK
                result:
                  message: Learner journey updated successfully
        '400':
          description: Bad Request
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.update
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: BAD_REQUEST
                result: null
                error:
                  code: LEARNER_JOURNEY_INVALID_INPUT
                  message: status must be either in_progress or complete
        '404':
          description: Not Found
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.update
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: NOT_FOUND
                result: null
                error:
                  code: LEARNER_NOT_FOUND
                  message: Learner with id 10 does not exist
  /v1/learner/journey/read/{learner_id}:
    get:
      tags:
        - default
      summary: Read
      security:
        - bearerAuth: []
      parameters:
        - name: msgid
          in: query
          schema:
            type: string
          example: d297e367-7b72-45c1-883c-18005a7ea888
        - name: learner_id
          in: path
          schema:
            type: string
          required: true
          example: 2ca3b11a-d581-4aa1-a199-fb1c51392cc9
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.fetch
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: SUCCESS
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: OK
                result:
                  identifier: d297e367-7b72-45c1-883c-18005a7ea888
                  learner_id: 2ca3b11a-d581-4aa1-a199-fb1c51392cc9
                  question_set_id: 37a60112-1136-40cc-b968-2f664feb1864
                  status: in_progress
                  attempts_count: 1
                  completed_question_ids:
                    - 123e4567-e89b-12d3-a456-426614174000
                  start_time: '2024-09-20T17:43:49+05:30'
                  end_time: null
                  created_at: '2024-09-20T17:43:49+05:30'
                  updated_at: '2024-09-20T17:43:49+05:30'
                  created_by: 123e4567-e89b-12d3-a456-426614174000
                  updated_by: 123e4567-e89b-12d3-a456-426614174000
        '400':
          description: Bad Request
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.fetch
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: BAD_REQUEST
                result: null
                error:
                  code: LEARNER_JOURNEY_INVALID_INPUT
                  message: learner_id must be an integer and is required
        '404':
          description: Not Found
          content:
            application/json:
              schema:
                type: object
              example:
                id: api.learner.journey.fetch
                ver: '1.0'
                ts: '2024-09-20T17:43:49+05:30'
                params:
                  status: FAILED
                  msgid: d297e367-7b72-45c1-883c-18005a7ea888
                  resmsgid: 01fc64f0-b525-4638-8668-2b6d8a26e481
                responseCode: NOT_FOUND
                result: null
                error:
                  code: LEARNER_NOT_FOUND
                  message: Learner with id 10 does not exist
```

File: src/controllers/LearnerJourneyUpdate/LearnerJourneyUpdate.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  const msgid = _.get(req, ['body', 'params', 'msgid']);
+  const dataBody = _.get(req, 'body.request');
+  const resmsgid = _.get(res, 'resmsgid');
+
+  const isRequestValid: Record<string, any> = schemaValidation(requestBody, learnerUpdateJSON);
+
+  const invalidStartTime = dataBody.start_time && !moment(dataBody.start_time).isValid();
+  const invalidEndTime = dataBody.end_time && !moment(dataBody.end_time).isValid();
+
+  const invalidDateTimeKey = invalidStartTime ? 'start_time' : 'end_time';
```
Comment (HarishGangula) [inline src/controllers/LearnerJourneyUpdate/LearnerJourneyUpdate.ts:25]: Add logic to check end time should be greater than start time
Check all the questions are attempted in question set

**Review by HarishGangula — APPROVED** [review]
Work on the TODO items

**Review by HarishGangula — APPROVED** [review]

**Pull request merged**